README.md
# Ceros Animation Plugin for WordPress

Unleash interactive creativity on your website with Ceros Embed. Elevate user engagement and infuse arts, music, and stunning visuals into your brand's digital presence.

# Table of Contents

- [Installation](#installation)
- [Usage](#usage)
- [Documentation](#documentation)
- [Support](#support)
- [Contributing](#contributing)
- [License](#license)

---

## Installation

1. Download the plugin zip file.
2. Log in to your WordPress dashboard.
3. Navigate to **Plugins** > **Add New**.
4. Click on the **Upload Plugin** button.
5. Choose the downloaded zip file and click **Install Now**.
6. After installation, click **Activate** to activate the Ceros Embed plugin.

## Usage

When adding your Ceros animation, upload a .zip file containing the Ceros project assets. Make sure that your files are at the root of the zipped folder. The plugin will handle the rest!

Please note that the plugin accepts .zip files. You can upload your project using and get the shortcode to embed it into any post, page, or custom post type.

To embed a Ceros project, use the following shortcode:

```
[ceros-animation id="THE_POST_ID" type="TYPE" title="YOUR_TITLE"]
```

Replace `THE_POST_ID` with the unique ID of your Ceros embed, `TYPE` with either "scrolling" or "full" (for different display types), and `YOUR_TITLE` with the title for screen readers.

## Documentation

For detailed information and guidelines, you can visit our [plugin documentation](https://prolificdigital.notion.site/Ceros-WordPress-Plugin-2165805d1e9f4fd5ad58ab64b5a5d473?pvs=4).

## Support

For support, feedback, or feature requests, please [submit an issue](https://github.com/prolific-digital/ceros-embed/issues) on GitHub.

Let your website come alive with the immersive power of Ceros Embed!

## Contributing

We welcome contributions to improve the WP Disable Comments plugin. To contribute:

1. Fork the project.
2. Make your feature addition or bug fix.
3. Send us a pull request with a brief description of your changes.

> Before sending a pull request for a feature addition, please discuss it with us first. We want to ensure that changes align with the project's direction and goals.

## License

This project is licensed under the GPL-3.0 License.

---

_Proudly crafted by Prolific Digital_


changelog.md
All notable changes to the "Ceros Animations" plugin will be documented in this file.

The format is based on Keep a Changelog and this project adheres to Semantic Versioning.

# Changelog

## Version 1.0.0 (08-07-2023)

- Initial release of the Ceros Animation Plugin for WordPress.
- Two display types: "scrolling" and "full" for flexible content presentation.
- Intuitive meta box for Ceros Custom Post Type management.
- Upload .zip files containing Ceros project assets for easy integration.
- Comprehensive documentation to guide users in setting up and using the plugin.
- Support for screen readers with customizable title attributes for enhanced accessibility.
- Seamless integration with WordPress, ensuring smooth installation and activation.
- Embed Ceros projects into posts, pages, or custom post types using the provided shortcode.


classes/CerosPostType.php
<?php

namespace CerosEmbed;

/**
 * Class CerosPostType
 */
class CerosPostType {

  /**
   * Initialize the shortcode and hooks
   */
  public function __construct() {
    add_action('init', array($this, 'custom_post_type_ceros'), 0);
  }

  /**
   * Register Custom Post Type: Ceros
   */
  public function custom_post_type_ceros() {
    $labels = array(
      'name' => _x('Ceros', 'Post Type General Name', 'your-plugin-textdomain'),
      'singular_name' => _x('Ceros', 'Post Type Singular Name', 'your-plugin-textdomain'),
      'menu_name' => __('Ceros', 'your-plugin-textdomain'),
      'name_admin_bar' => __('Ceros', 'your-plugin-textdomain'),
      'archives' => __('Ceros Archives', 'your-plugin-textdomain'),
      'attributes' => __('Ceros Attributes', 'your-plugin-textdomain'),
      'parent_item_colon' => __('Parent Ceros:', 'your-plugin-textdomain'),
      'all_items' => __('All Ceros', 'your-plugin-textdomain'),
      'add_new_item' => __('Add New Ceros', 'your-plugin-textdomain'),
      'add_new' => __('Add New', 'your-plugin-textdomain'),
      'new_item' => __('New Ceros', 'your-plugin-textdomain'),
      'edit_item' => __('Edit Ceros', 'your-plugin-textdomain'),
      'update_item' => __('Update Ceros', 'your-plugin-textdomain'),
      'view_item' => __('View Ceros', 'your-plugin-textdomain'),
      'view_items' => __('View Ceros', 'your-plugin-textdomain'),
      'search_items' => __('Search Ceros', 'your-plugin-textdomain'),
      'not_found' => __('Not found', 'your-plugin-textdomain'),
      'not_found_in_trash' => __('Not found in Trash', 'your-plugin-textdomain'),
      'featured_image' => __('Featured Image', 'your-plugin-textdomain'),
      'set_featured_image' => __('Set featured image', 'your-plugin-textdomain'),
      'remove_featured_image' => __('Remove featured image', 'your-plugin-textdomain'),
      'use_featured_image' => __('Use as featured image', 'your-plugin-textdomain'),
      'insert_into_item' => __('Insert into Ceros', 'your-plugin-textdomain'),
      'uploaded_to_this_item' => __('Uploaded to this Ceros', 'your-plugin-textdomain'),
      'items_list' => __('Ceros list', 'your-plugin-textdomain'),
      'items_list_navigation' => __('Ceros list navigation', 'your-plugin-textdomain'),
      'filter_items_list' => __('Filter Ceros list', 'your-plugin-textdomain'),
    );

    $args = array(
      'label' => __('Ceros', 'your-plugin-textdomain'),
      'description' => __('Ceros created with your plugin.', 'your-plugin-textdomain'),
      'labels' => $labels,
      'supports' => array('title'),
      'taxonomies' => array(),
      'hierarchical' => false,
      'public' => false,
      'show_ui' => true,
      'show_in_menu' => true,
      'menu_position' => 5,
      'menu_icon' => 'dashicons-images-alt2',
      'show_in_admin_bar' => false,
      'show_in_nav_menus' => false,
      'can_export' => true,
      'has_archive' => false,
      'exclude_from_search' => true,
      'publicly_queryable' => false,
      'capability_type' => 'post',
    );

    register_post_type('ceros', $args);
  }
}


classes/CerosShortcode.php
<?php

namespace CerosEmbed;

/**
 * Class CerosShortcode
 */
class CerosShortcode {

  /**
   * Initialize the shortcode and hooks
   */
  public function __construct() {
    add_shortcode('ceros-animation', array($this, 'render_shortcode'));
  }

  /**
   * Render the Ceros project shortcode
   *
   * @param array $atts Shortcode attributes.
   * @return string
   */
  public function render_shortcode($atts) {
    // Shortcode attributes
    $atts = shortcode_atts(
      array(
        'id' => '',
        'type' => 'scrolling',
        'title' => '',
      ),
      $atts,
      'ceros-animation'
    );

    // Extract shortcode attributes
    $project_id = $atts['id'];
    $type = $atts['type'];
    $title = $atts['title'];
    $site_url = get_site_url();

    // Check if a project ID is provided
    if (empty($project_id)) {
      return '<p style="color: #f00;">Error: Ceros project ID is missing. Please provide a valid project ID.</p>';
    }

    // Get the public URL to the project folder within the 'embeds' directory
    $project_url = plugin_dir_url(__FILE__) . 'embeds/' . sanitize_file_name($project_id);

    // Embed Ceros project using the public URL
    $file_url = get_post_meta($project_id, '_ceros_file_upload', true);

    if ($type == 'scrolling') {
      // Scrolling
      $embed_code = '<div style="position: relative;width: auto;padding: 0 0 56.25%;height: 0;top: 0;left: 0;bottom: 0;right: 0;margin: 0;border: 0 none" id="experience-' . uniqid() . '" data-aspectRatio="1.77777778" data-mobile-aspectRatio="0.5625" data-tablet-aspectRatio="0.75"><iframe allowfullscreen src="' . esc_url($file_url) . '/index.html" style="position: absolute;top: 0;left: 0;bottom: 0;right: 0;margin: 0;padding: 0;border: 0 none;height: 100%;width: 100%" frameborder="0" class="ceros-experience" title="' . $title . '"></iframe></div><script type="text/javascript" src="survey-report/assets/scroll-proxy.min.js" data-ceros-origin-domains="' . $site_url . '"></script>';
    } else if ($type == 'full') {
      // Full height
      $embed_code = '<div style="position: relative;width: auto;padding: 0 0 432.73%;height: 0;top: 0;left: 0;bottom: 0;right: 0;margin: 0;border: 0 none" id="experience-' . uniqid() . '" data-aspectRatio="0.23108864" data-mobile-aspectRatio="0.08587786" data-tablet-aspectRatio="0.1221374"><iframe allowfullscreen src="' . esc_url($file_url) . '?heightOverride=5539&mobileHeightOverride=6288&tabletHeightOverride=6288" style="position: absolute;top: 0;left: 0;bottom: 0;right: 0;margin: 0;padding: 0;border: 0 none;height: 1px;width: 1px;min-height: 100%;min-width: 100%" frameborder="0" class="ceros-experience" title="' . $title . '" scrolling="no"></iframe></div><script type="text/javascript" src="survey-report/assets/scroll-proxy.min.js" data-ceros-origin-domains="' . $site_url . '"></script>';
    }

    return $embed_code;
  }
}


classes/CerosPostOptions.php
<?php

namespace CerosEmbed;

/**
 * Class CerosPostOptions
 */
class CerosPostOptions {

  /**
   * Initialize the shortcode and hooks
   */
  public function __construct() {
    add_action('add_meta_boxes', array($this, 'add_meta_box'));
    add_action('add_meta_boxes', array($this, 'add_meta_box_sidebar'));
    add_filter('upload_mimes', array(
      $this, 'custom_upload_mimes'
    ));
    add_action('save_post_ceros', array($this, 'handle_upload'));
  }

  // Add the custom meta box for the "Ceros" custom post type
  public function add_meta_box() {
    add_meta_box(
      'ceros_file_upload', // Meta box ID
      'Ceros Animation', // Meta box title
      array(
        $this, 'render_meta_box'
      ), // Callback to render the meta box content
      'ceros', // Custom post type slug
      'normal', // Context (normal, side, advanced)
      'default' // Priority (default, high, low)
    );
  }

  // Add the custom meta box for the "Ceros" custom post type
  public function add_meta_box_sidebar() {
    add_meta_box(
      'ceros_file_upload_d', // Meta box ID
      'Ceros Animation', // Meta box title
      array(
        $this, 'side_content'
      ), // Callback to render the meta box content
      'ceros', // Custom post type slug
      'side', // Context (normal, side, advanced)
      'default' // Priority (default, high, low)
    );
  }

  public function side_content() {
    echo '<p><b>ID:</b> unique ID for this embed</p>';
    echo '<p><b>Type:</b> scroll, full</p>';
    echo '<p><b>Title:</b> used for screen readers</p>';
    echo '<p><a href="https://prolificdigital.notion.site/Ceros-WordPress-Plugin-2165805d1e9f4fd5ad58ab64b5a5d473?pvs=4" target="_blank">View documentation for details</a></p>';
  }

  // Callback to render the custom meta box content
  public function render_meta_box($post) {
    wp_nonce_field(basename(__FILE__), 'ceros_file_upload_nonce');
    $file_url = get_post_meta($post->ID, '_ceros_file_upload', true);
?>

    <input type="file" id="ceros_file_upload_field" name="ceros_file_upload" accept=".zip">
    <p><i>Upload your .zip file above and be sure that your files are at the root of your zipped folder. </br></i><a href="https://prolificdigital.notion.site/Ceros-WordPress-Plugin-2165805d1e9f4fd5ad58ab64b5a5d473?pvs=4" target="_blank">View our documentation</a> for more details.</p>

    <p><?php echo $file_url ? '<p><b>Current File: </b></p>' . $file_url . '/index.html' : '<p><b>No file uploaded yet.</b></p>'; ?></p>

    <p><b>Shortcode:</b></p>
    <pre><input type="text" readonly value="[ceros-animation id=&quot;<?php the_ID(); ?>&quot; type=&quot;scrolling&quot; title=&quot;<?php the_title(); ?>&quot;]" class="widefat"></pre>

    <?php if ($file_url) : ?>
      <p><b>Preview:</b></p>
      <div style='position: relative;width: auto;padding: 0 0 56.25%;height: 0;top: 0;left: 0;bottom: 0;right: 0;margin: 0;border: 0 none' id="experience-5fa993f346f86" data-aspectRatio="1.77777778" data-mobile-aspectRatio="0.5625" data-tablet-aspectRatio="0.75"><iframe allowfullscreen src='<?php echo $file_url; ?>/index.html' style='position: absolute;top: 0;left: 0;bottom: 0;right: 0;margin: 0;padding: 0;border: 0 none;height: 100%;width: 100%' frameborder='0' class='ceros-experience' title='2020 Health Plan Member Engagement Report'></iframe></div>
      <script type="text/javascript" src="survey-report/assets/scroll-proxy.min.js" data-ceros-origin-domains=""></script>
    <?php endif; ?>

    <script>
      // This is neccessary to handle file uploads.
      jQuery(document).ready(function() {
        jQuery('#post').attr('enctype', 'multipart/form-data');
      });
    </script>
<?php
  }

  public function custom_upload_mimes($mime_types) {
    return array_merge($mime_types, array(
      'zip' => 'application/zip',
      // Need to test future support for these file types
      // 'tar' => 'application/x-tar',
      // 'tgz' => 'application/x-tar-gz',
      // 'gzip' => 'application/x-zip',
      // 'gz' => 'application/x-zip',
    ));
    return $mime_types;
  }

  public function handle_upload($post_id) {
    if (!isset($_POST['ceros_file_upload_nonce']) || !wp_verify_nonce($_POST['ceros_file_upload_nonce'], basename(__FILE__))) {
      return;
    }
    if (defined('DOING_AUTOSAVE') && DOING_AUTOSAVE) {
      return;
    }
    if (!current_user_can('edit_post', $post_id)) {
      return;
    }

    $file = $_FILES['ceros_file_upload'];

    if ($file['error'] === 0) {
      $source = $_FILES['ceros_file_upload']['tmp_name'];

      // Set the custom uploads directory path
      $upload_dir = wp_upload_dir();
      wp_mkdir_p($upload_dir['basedir'] . '/ceros');
      $custom_directory = $upload_dir['basedir'] . '/ceros';

      $filename = sanitize_file_name($_FILES['ceros_file_upload']['name']);
      $base_filename = pathinfo($filename, PATHINFO_FILENAME);
      $destination = trailingslashit($custom_directory) . $filename;

      // Increment the ZIP file name if one with the current name already exists
      if (file_exists($destination)) {
        $extension = pathinfo($filename, PATHINFO_EXTENSION);
        $base_filename = pathinfo($filename, PATHINFO_FILENAME);
        $i = 1;
        while (file_exists($destination)) {
          $filename = $base_filename . '_' . $i . '.' . $extension;
          $destination = trailingslashit($custom_directory) . $filename;
          $i++;
        }
      }

      move_uploaded_file($source, $destination);

      $folder_name = pathinfo($filename, PATHINFO_FILENAME);
      $unzip_folder = trailingslashit($custom_directory) . $folder_name;

      // Create a temp folder to initially unzip the files to
      $temp_unzip_folder = trailingslashit($custom_directory) . 'temp_unzip';
      wp_mkdir_p($temp_unzip_folder);

      WP_Filesystem();
      $unzipped_file = unzip_file($destination, $temp_unzip_folder);

      // If there's a single directory inside temp_unzip_folder, move its contents to the desired destination
      $dir_contents = scandir($temp_unzip_folder);
      if (count($dir_contents) == 3) { // "." , "..", and the directory
        $nested_folder = trailingslashit($temp_unzip_folder) . $dir_contents[2]; // The directory
        rename($nested_folder, $unzip_folder);
      }

      // Remove the temp folder after the operations
      rmdir($temp_unzip_folder);

      if ($unzipped_file) {
        $file_path = $upload_dir['baseurl'] . '/ceros/' . $folder_name;
        update_post_meta($post_id, '_ceros_file_upload', $file_path);
      }
    }
  }
}
